File: backend/src/learning/stage.ts
/**
 * The single source of truth for OpenMind's stage-based product rule:
 * one app, one identity, one tutor — two stage-appropriate learning
 * experiences. Everything that branches on the learner's stage (routes,
 * tutor prompts, views, the Flutter twin core/stage.dart) derives from the
 * authenticated grade through THIS module, never from ad-hoc comparisons.
 */

export const MIN_GRADE = 1;
export const MAX_GRADE = 9;

/** Last grade of the elementary games product. */
export const PRIMARY_MAX_GRADE = 6;

export const LEARNING_STAGES = ['primary_games', 'middle_interactive_learning'] as const;
export type LearningStage = (typeof LEARNING_STAGES)[number];

/** Grades 1-6 → the elementary games product; 7-9 → interactive learning. */
export function stageForGrade(grade: number): LearningStage {
  return grade <= PRIMARY_MAX_GRADE ? 'primary_games' : 'middle_interactive_learning';
}

/**
 * The elementary game-generation pipeline (prompts, validators, fact-check)
 * is calibrated for grades 1-6 ONLY. A middle-school student may still open
 * legacy games, so any grade fed into that pipeline is clamped HERE — at the
 * generation boundary — and nowhere else. Student identity keeps the true grade.
 */
export function gameGenGrade(grade: number): number {
  return Math.min(Math.max(grade, MIN_GRADE), PRIMARY_MAX_GRADE);
}

/**
 * Middle-school context lenses ("عدسة التعلم") — a lightweight learner
 * preference that flavors examples and framing. Deliberately separate from
 * the elementary Student.interest game archetypes. Ids are stable API values;
 * display strings live in the Flutter localization layer.
 */
export const LEARNING_CONTEXTS = [
  'market', // السوق والتجارة
  'building', // البناء والعمران
  'water_energy', // الماء والطاقة
  'roads_transport', // الطرق والمواصلات
  'technology', // التقنية والاتصالات
] as const;
export type LearningContext = (typeof LEARNING_CONTEXTS)[number];

File: edumind-ui/lib/core/stage.dart
/// The stage-based product rule — Dart twin of backend/src/learning/stage.ts.
///
/// One OpenMind app, two stage-appropriate learning experiences:
///  - grades 1-6  → [LearningStage.primaryGames] (the elementary games product)
///  - grades 7-9  → [LearningStage.middleInteractiveLearning] (journeys/tutor)
///
/// Every screen, route, or service that adapts to the learner's stage resolves
/// it HERE (usually via [Session.stage]) — never with its own grade math.
/// The backend remains the trusted source: the resolved `stage` string it
/// returns on the student view is cached in the profile and preferred over
/// the local grade fallback.
library;

enum LearningStage {
  primaryGames,
  middleInteractiveLearning;

  /// Wire value — matches the backend's LEARNING_STAGES exactly.
  String get wire => switch (this) {
        LearningStage.primaryGames => 'primary_games',
        LearningStage.middleInteractiveLearning => 'middle_interactive_learning',
      };

  static LearningStage? fromWire(String? v) => switch (v) {
        'primary_games' => LearningStage.primaryGames,
        'middle_interactive_learning' => LearningStage.middleInteractiveLearning,
        _ => null,
      };
}

const int kMinGrade = 1;
const int kMaxGrade = 9;

/// Last grade of the elementary games product.
const int kPrimaryMaxGrade = 6;

/// Grades 1-6 → primary games; 7-9 → middle-school interactive learning.
/// Offline fallback only — the backend-resolved stage wins when cached.
LearningStage stageForGrade(int grade) => grade <= kPrimaryMaxGrade
    ? LearningStage.primaryGames
    : LearningStage.middleInteractiveLearning;

/// Middle-school context lenses — ids match the backend's LEARNING_CONTEXTS.
/// A lens flavors story framing and examples; it never changes the concept,
/// difficulty, targets, or progress. Labels live in AppLocalizations
/// (`ctx_<id>`); emoji here so data and UI stay in one place.
const List<({String id, String emoji})> kLearningContexts = [
  (id: 'market', emoji: '🛒'),
  (id: 'building', emoji: '🏗️'),
  (id: 'water_energy', emoji: '💧'),
  (id: 'roads_transport', emoji: '🚌'),
  (id: 'technology', emoji: '📱'),
];

bool isSupportedLearningContext(String id) =>
    kLearningContexts.any((c) => c.id == id);
